scripts/imgtool/main.py

```diff
@@ -477,10 +502,45 @@ def sign(key, public_key_format, align, version, pad_sig, header_size,
         }
 
     img.create(key, public_key_format, enckey, dependencies, boot_record,
-               custom_tlvs, int(encrypt_keylen), clear, baked_signature,
+               custom_tlvs, compression_tlvs, int(encrypt_keylen), clear, baked_signature,
                pub_key, vector_to_sign)
-    img.save(outfile, hex_addr)
 
+    if os.path.splitext(infile)[-1].lower() == '.bin' :
+        if compression == "lzma2" :
+            compressed_img = image.Image(version=decode_version(version), header_size=header_size,
+                      pad_header=pad_header, pad=pad, confirm=confirm,
+                      align=int(align), slot_size=slot_size,
+                      max_sectors=max_sectors, overwrite_only=overwrite_only,
+                      endian=endian, load_addr=load_addr, rom_fixed=rom_fixed,
+                      erased_val=erased_val, save_enctlv=save_enctlv,
+                      security_counter=security_counter, max_align=max_align)
+            compressed_file =  re.sub('zephyr\.bin$','zephyr.compressed.bin',infile)
+            compression_filters = [
+                {"id": lzma.FILTER_LZMA2, "preset": comp_default_preset, "dict_size": comp_default_dictsize, "lp": comp_default_lp, "lc": comp_default_lc}
+            ]
+            in_fd=open(infile, "rb")
+            with lzma.open(compressed_file, "wb", filters=compression_filters, format=lzma.FORMAT_RAW) as out_fd:
+                out_fd.write(in_fd.read())
```

Comment:
don't think we need to output the compressed contents separately, we don't do the same thing with encrypted data

Reply:
Compress in memory - see https://docs.python.org/3/library/lzma.html#compressing-and-decompressing-data-in-memory